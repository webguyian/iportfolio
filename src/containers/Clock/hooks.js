import { useEffect, useRef, useState } from 'react';
import { useHistory } from 'react-router-dom';

import { initialValues } from './constants';
import { getExpiration, getExpirationValues, getSeconds } from './helpers';

export const useInitialRoute = (match, path) => {
  const history = useHistory();

  useEffect(() => {
    if (match.isExact) {
      history.push(path);
    }
  }, []);

  return history;
};

export const useStorage = (key, currentValue, deleteKey, keepUpdating) => {
  const getValue = () => {
    try {
      // Get item from local storage with key
      const item = window.localStorage.getItem(key);

      // Parse stored JSON or if none return undefined
      return item ? JSON.parse(item) : undefined;
    } catch (error) {
      // If error return undefined
      return undefined;
    }
  };

  const value = getValue();
  const localRef = useRef(value);
  const [storedValue, setStoredValue] = useState(value);

  const setValue = nextValue => {
    try {
      // Save to state and local storage
      setStoredValue(nextValue);

      if (deleteKey && !nextValue[deleteKey]) {
        // Remove item from local storage
        window.localStorage.removeItem(key);
      } else {
        // Update local storage and include timestamp
        nextValue.timestamp = Date.now();
        window.localStorage.setItem(key, JSON.stringify(nextValue));
      }
    } catch (error) {
      // Drop error
    }
  };

  useEffect(() => {
    // Update local value
    localRef.current = currentValue || storedValue;

    if (keepUpdating) {
      // Set update value in local storage
      setValue(localRef.current);
    }
  }, [currentValue, storedValue]);

  useEffect(() => {
    return () => {
      if (key && localRef.current) {
        setValue(localRef.current);
      }
    };
  }, []);

  return localRef.current;
};

export const useLaps = (initialLaps, timer, timerRunning) => {
  const [laps, setLaps] = useState(initialLaps);
  const hasLaps = laps.some(lap => lap > 0);
  const storage = useStorage(
    'stopwatch',
    {
      laps,
      timer,
      timerRunning
    },
    'timer'
  );

  const updateLaps = () => {
    if (timerRunning) {
      const hasEmptyLaps = laps.some(lap => lap === 0);

      if (hasEmptyLaps) {
        setLaps(updatedLaps => [
          0,
          ...updatedLaps.slice(0, updatedLaps.length - 1)
        ]);
      } else {
        setLaps(updatedLaps => [0, ...updatedLaps]);
      }
    } else {
      setLaps(initialLaps);
    }
  };

  useEffect(() => {
    if (storage && storage.laps) {
      setLaps(storage.laps);
    }
  }, []);

  useEffect(() => {
    if (!timerRunning || !timer) {
      return;
    }

    const lapLength = laps.length;

    if (!hasLaps && timer > 10) {
      setLaps(updatedLaps => [timer, ...updatedLaps]);
    }

    if (lapLength > 5) {
      setLaps(updatedLaps => {
        const truncatedLaps = updatedLaps.slice(1, lapLength);
        const offset = truncatedLaps.reduce((a, b) => a + b, 0);

        return [timer - offset, ...truncatedLaps];
      });
    }
  }, [timer, timerRunning, hasLaps]);

  return [laps, updateLaps, hasLaps];
};

export const useStopwatch = (startTime = 0, start = false) => {
  const [timerStarted, startTimer] = useState(start);
  const [interval, setLocalInterval] = useState(null);
  const [offset, setOffset] = useState(0);
  const timer = useRef(startTime);
  const storage = useStorage('stopwatch');
  const toggleTimer = () => startTimer(!timerStarted);
  const resetTimer = () => {
    timer.current = 0;
    return timer.current;
  };
  const getUpdatedTime = cache => {
    const now = Date.now();
    const diff = now - cache.timestamp;

    return cache.timer + diff;
  };

  useEffect(() => {
    if (storage && storage.timer) {
      // Get updated time from local storage
      if (storage.timerRunning) {
        timer.current = getUpdatedTime(storage);
        toggleTimer();
      } else {
        timer.current = storage.timer;
      }
    }
  }, []);

  useEffect(() => {
    let localOffset = offset;

    if (timerStarted && !interval) {
      localOffset = Date.now();

      const getDelta = () => {
        const now = Date.now();
        const delta = now - localOffset;

        localOffset = now;

        setOffset(now);
        return delta;
      };

      const localInterval = window.setInterval(() => {
        timer.current += getDelta();
      }, 1);

      setLocalInterval(localInterval);
    }

    if (!timerStarted && interval) {
      window.clearInterval(interval);
      setLocalInterval(null);
    }
  }, [timerStarted, timer, interval]);

  return [timer.current, timerStarted, toggleTimer, resetTimer];
};

export const useDuration = () => {
  const [values, setValues] = useState(null);
  const storage = useStorage('duration', values, null, true);
  const setDuration = (key, value) => {
    setValues({
      ...values,
      [key]: value
    });
  };
  const cachedValues = storage || values;
  const duration = cachedValues && getSeconds(cachedValues);

  useEffect(() => {
    if (storage) {
      // Update state with values from local storage
      setValues(storage);
    } else {
      // Set default initial values
      setValues(initialValues);
    }
  }, []);

  return [duration, setDuration, cachedValues, setValues];
};

export const useCountdown = (h = 0, m = 0, s = 0) => {
  const [seconds, setSeconds] = useState(s);
  const [minutes, setMinutes] = useState(m);
  const [hours, setHours] = useState(h);
  const allSeconds = getSeconds({ hours, minutes, seconds });

  const subtractHour = () => {
    setHours(prevHours => {
      if (prevHours === 0) {
        return 23;
      } else if (prevHours > 0) {
        return prevHours - 1;
      }
      return 0;
    });
  };

  const subtractMinute = () => {
    setMinutes(prevMinutes => {
      if (prevMinutes === 0) {
        subtractHour();
        return 59;
      } else if (prevMinutes > 0) {
        return prevMinutes - 1;
      }
      return 0;
    });
  };

  const subtractSecond = () => {
    setSeconds(prevSeconds => {
      if (prevSeconds === 0) {
        subtractMinute();
        return 59;
      } else if (prevSeconds > 0) {
        return prevSeconds - 1;
      }
      return 0;
    });
  };

  const setCountdown = values => {
    if (!values) {
      setHours(0);
      setMinutes(0);
      setSeconds(0);
      return;
    }

    setHours(values.hours);
    setMinutes(values.minutes);
    setSeconds(values.seconds);
  };

  return [
    {
      allSeconds,
      hours,
      minutes,
      seconds
    },
    setCountdown,
    subtractSecond
  ];
};

export const useTimer = duration => {
  const [started, setStarted] = useState(false);
  const [running, setRunning] = useState(false);
  const [countdown, setCountdown, startCountdown] = useCountdown();
  const intervalRef = useRef();
  const storage = useStorage(
    'timer',
    {
      ...countdown,
      duration,
      running,
      started
    },
    'allSeconds'
  );

  const pause = () => {
    if (intervalRef.current) {
      setRunning(() => false);
      clearInterval(intervalRef.current);
      intervalRef.current = undefined;
    }
  };

  const reset = () => {
    pause();
    setCountdown(null);
    setStarted(false);
  };

  const calculateExpiration = expiration => {
    const distance = expiration - Date.now();
    const expirationValues = getExpirationValues(distance);

    if (countdown.allSeconds < 0) {
      reset();
    } else {
      setCountdown(expirationValues);
    }

    return expirationValues;
  };

  const start = expiration => {
    if (expiration && !intervalRef.current) {
      calculateExpiration(expiration);
    }
  };

  const resume = () => {
    if (!intervalRef.current) {
      setStarted(() => true);
      setRunning(() => true);
      intervalRef.current = setInterval(startCountdown, 1000);
    }
  };

  const restart = () => {
    const expiration = getExpiration(duration);

    reset();
    start(expiration);
  };

  const toggle = () => {
    if (!intervalRef.current) {
      resume();
    } else {
      pause();
    }
  };

  useEffect(() => {
    const expiration = getExpiration(duration, storage);

    start(expiration);

    if (storage) {
      if (storage.running) {
        resume();
      } else if (storage.started) {
        setStarted(true);
      }

      if (duration !== storage.duration) {
        // Remove timer from local storage
        window.localStorage.removeItem('timer');
      }
    }

    return pause;
  }, [duration]);

  return {
    ...countdown,
    duration,
    start,
    started,
    pause,
    reset,
    resume,
    restart,
    running,
    toggle
  };
};

export const useRadialOffset = (timer, radius) => {
  const circumference = (2 * radius * 22) / 7;
  const { allSeconds, duration } = timer;
  const [initialSeconds] = useState(allSeconds);
  const [offset, setOffset] = useState(0);

  useEffect(() => {
    if (!offset) {
      const initialOffset = (allSeconds / duration) * circumference;
      const updatedOffset = circumference - Math.round(initialOffset);

      if (updatedOffset === circumference || updatedOffset < 0) {
        // Exit early
        return;
      }

      setOffset(updatedOffset);
    }
  }, [allSeconds, duration, offset]);

  return [offset, initialSeconds];
};
