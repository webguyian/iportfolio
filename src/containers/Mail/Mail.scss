.mail-app {
  @include stretch-column-nowrap;
  position: relative;
  top: $gutter;
  transition: top 0.4s;
  border-top-left-radius: 10px;
  border-top-right-radius: 10px;
  background-color: $black-dark;
  text-align: left;
  animation: slide-up 0.4s linear;

  &-drag-handle {
    display: block;
    width: 40px;
    margin: $gutter-s auto;
    padding: 3px;
    border-radius: $border-radius-m;
    background-color: $gray-medium;
  }

  &-top-bar {
    padding: 0 $gutter;

    > .ui-link {
      color: $blue;
    }
  }

  &-header {
    padding: $gutter 0 8px $gutter;

    &-title {
      display: flex;
      justify-content: space-between;
      margin-bottom: $gutter;

      > .ui-text--display {
        word-break: break-word;
      }

      > .ui-btn--with-icon {
        @include circle(36px);
        flex-shrink: 0;
        margin-right: $gutter;
        background-color: $blue;

        &[disabled] {
          background-color: $gray-medium;
          color: $black-dark;
          cursor: default;
        }
      }
    }

    .ui-form-field {
      padding-left: 0;
      border-bottom: 1px solid $gray-dark;

      &-label {
        color: $gray;
      }

      &-input[disabled] {
        color: $blue;
      }
    }
  }

  &-body {
    height: 100%;
    max-height: 515px;
    overflow-y: auto;
  }

  &-overlay {
    position: fixed;
    left: 28px;
    width: 375px;
    height: 760px;
    background: linear-gradient(180deg, rgba($black-dark, 0.8) 55%, rgba($black-dark, 0.5) 95%, rgba($black-dark, 0) 100%);
  }

  &-controls {
    position: relative;
    bottom: 30px;
    padding: 0 8px;
    transform: translateY(110%);
    transition: transform 0.4s;

    &--show {
      transform: translateY(0);
    }

    > .ui-btn--anchor-block {
      transition: background-color 0.2s;
      background-color: $gray-dark;
      color: $blue;

      &:hover,
      &:focus {
        background-color: darken($gray-dark, 2%);
      }

      &:first-child {
        border-bottom: 1px solid $black-dark;
        border-top-left-radius: $border-radius-l;
        border-top-right-radius: $border-radius-l;
        color: $red;
      }

      &:nth-child(2) {
        border-bottom-left-radius: $border-radius-l;
        border-bottom-right-radius: $border-radius-l;
      }

      &:last-child {
        margin-top: 8px;
        border-radius: $border-radius-l;
      }
    }
  }
}

@keyframes slide-up {
  from {
    top: 550px;
  }
}
