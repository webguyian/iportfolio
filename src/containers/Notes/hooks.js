import { useEffect, useRef, useState } from 'react';
import { useSwipeable } from 'react-swipeable';

export const useRefFocus = (condition = true, initial = null) => {
  const ref = useRef(initial);

  useEffect(() => {
    const element = ref && ref.current;

    if (condition && element) {
      // Add focus to element
      element.focus();
    }
  }, [condition, ref]);

  return ref;
};

export const useSwipeOffset = (initial = 0) => {
  const [offset, setOffset] = useState(initial);
  const maxOffset = -30;

  const handleSwipe = eventInfo => {
    if (eventInfo.dir !== 'Down') {
      // Exit early if not down swipe
      return false;
    }

    setOffset(eventInfo.deltaY);
  };

  const handleSwipeUp = () => {
    if (!offset) {
      // Exit early with no offset
      return false;
    }

    setOffset(0);
  };

  const handlers = useSwipeable({
    onSwipedUp: handleSwipeUp,
    onSwiping: handleSwipe,
    trackMouse: true
  });

  return [handlers, offset < maxOffset, setOffset];
};
