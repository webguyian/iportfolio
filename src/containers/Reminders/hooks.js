import { useEffect, useState, useRef } from 'react';
import { useSwipeable } from 'react-swipeable';

export const useLocalStorage = (key, initialValue) => {
  const getValue = () => {
    try {
      // Get item from local storage with key
      const item = window.localStorage.getItem(key);

      // Parse stored json or if none return initialValue
      return item ? JSON.parse(item) : initialValue;
    } catch (error) {
      // If error return initialValue
      return initialValue;
    }
  };
  const [storedValue, setStoredValue] = useState(getValue);
  const setValue = value => {
    try {
      // Allow value to be a function so we have same API as useState
      const valueToStore =
        value instanceof Function ? value(storedValue) : value;

      // Save state
      setStoredValue(valueToStore);
      // Save to local storage
      window.localStorage.setItem(key, JSON.stringify(valueToStore));
    } catch (error) {
      // Drop error
    }
  };

  return [storedValue, setValue, getValue];
};

export const useRefFocus = (node = null) => {
  const ref = useRef(node);

  useEffect(() => {
    const element = ref && ref.current;

    if (element) {
      // Add focus to element
      element.focus();
    }
  }, [ref]);

  return ref;
};

export const useRefControlledFocus = (focused, node = null) => {
  const ref = useRef(node);

  useEffect(() => {
    const element = ref && ref.current;

    if (element && focused) {
      element.focus();
    }
  }, [focused]);

  return ref;
};

export const useReminders = (reminders, setReminders) => {
  const remindersRef = useRef(reminders);

  useEffect(() => {
    // Update reminders in ref
    remindersRef.current = reminders;
  });

  useEffect(() => {
    return () => {
      const { current } = remindersRef;
      const updatedReminders = current.filter(reminder =>
        reminder.value.trim()
      );

      // Trigger callback without empty reminders
      setReminders(updatedReminders);
    };
  }, [remindersRef]);

  return remindersRef;
};

export const useSwipeOffset = (callback, initial = 0) => {
  const [offset, setOffset] = useState(initial);

  const handleSwipe = eventInfo => {
    if (eventInfo.dir !== 'Left') {
      // Exit early if not left swipe
      return false;
    }

    setOffset(eventInfo.deltaX);
  };

  const handleSwipeLeft = eventInfo => {
    const { deltaX } = eventInfo;

    if (deltaX > 80) {
      callback(eventInfo);
    }
  };

  const handleSwipeRight = () => {
    if (!offset) {
      // Exit early with no offset
      return false;
    }

    setOffset(0);
  };

  const handlers = useSwipeable({
    onSwipedLeft: handleSwipeLeft,
    onSwipedRight: handleSwipeRight,
    onSwiping: handleSwipe,
    trackMouse: true
  });

  return [handlers, offset, setOffset];
};
