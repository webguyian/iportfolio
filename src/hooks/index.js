import { useEffect, useState } from 'react';

export const useBreakpoint = breakpoint => {
  const [matchesBreakpoint, setBreakpoint] = useState(false);
  const mediaQuery = `(max-width: ${breakpoint}px)`;

  const handleResize = () => {
    const matches = window.matchMedia(mediaQuery).matches;

    setBreakpoint(matches);
  };

  useEffect(() => {
    window.addEventListener('resize', handleResize);

    return () => {
      window.removeEventListener('resize', handleResize);
    };
  }, [breakpoint]);

  return matchesBreakpoint;
};
