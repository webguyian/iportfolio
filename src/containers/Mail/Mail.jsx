import React from 'react';
import { object } from 'prop-types';
import classNames from 'classnames';

import Button from 'components/Button/Button';
import FormField from 'components/FormField/FormField';
import Link from 'components/Link/Link';
import Text from 'components/Text/Text';

import { useMail } from 'modules/mail/hooks';

const Mail = props => {
  const { location } = props;
  const [fields, setField, state, handlers] = useMail(location.state);
  const baseClass = 'mail-app';
  const headerClass = `${baseClass}-header`;
  const attachmentClass = `${baseClass}-attachment`;
  const controlsClass = `${baseClass}-controls`;
  const showControlsClass = state.showControls && `${controlsClass}--show`;
  const title = fields.subject || 'New Message';

  return (
    <form
      className={baseClass}
      onSubmit={handlers.onSubmit}
      {...handlers.swipe}
    >
      {state.showControls && <div className={`${baseClass}-overlay`} />}
      <Button
        className={`${baseClass}-drag-handle`}
        aria-labelledby={baseClass}
      >
        <Text type="accessible">Drag down to close</Text>
      </Button>
      <div className={`${baseClass}-top-bar`}>
        <Link to="/home" onClick={handlers.onCancel}>
          Cancel
        </Link>
      </div>
      <div className={headerClass}>
        <div className={`${headerClass}-title`}>
          <Text type="display" size="l">
            {title}
          </Text>
          <Button
            disabled={state.invalid}
            icon="arrow-up"
            size="2x"
            type="submit"
          >
            Send
          </Button>
        </div>
        <div className={`${headerClass}-fields`}>
          <FormField
            id="to"
            label="To:"
            disabled
            onChange={setField}
            value={fields.to}
          />
          <FormField
            id="from"
            label="From:"
            onChange={setField}
            placeholder="your@email.com"
            type="email"
            value={fields.from}
          />
          <FormField
            id="subject"
            label="Subject:"
            onChange={setField}
            value={fields.subject}
            maxLength={35}
          />
        </div>
      </div>
      <FormField
        className={`${baseClass}-body`}
        id="body"
        onChange={setField}
        type="textarea"
        value={fields.body}
      />
      <div className={attachmentClass}>
        {fields.attachment && (
          <img
            className={`${attachmentClass}-image`}
            alt="Attached image"
            src={fields.attachment}
          />
        )}
      </div>
      <div className={classNames(controlsClass, showControlsClass)}>
        <Button modifier="anchor-block" onClick={handlers.onDelete}>
          Delete Draft
        </Button>
        <Link to="/home" className="ui-btn--anchor-block">
          Save Draft
        </Link>
        <Button modifier="anchor-block" onClick={handlers.onConfirmCancel}>
          Cancel
        </Button>
      </div>
    </form>
  );
};

Mail.propTypes = {
  location: object.isRequired
};

export default Mail;
