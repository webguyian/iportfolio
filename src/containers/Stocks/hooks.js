import { useEffect, useState } from 'react';

import { useStorageCache } from 'hooks';
import { API_EXCHANGE, API_TOKEN } from 'containers/Stocks/constants';
import {
  getCandlestickEndpoint,
  getCandlestickEndpoints,
  getStockEndpoint,
  getStockNewsEndpoint,
  isExpired,
  isExpiredNews,
  updateChartData,
  rehydrateChartData
} from 'containers/Stocks/helpers';

export const useFetch = endpoint => {
  const [data, setData] = useState(null);
  const fetchData = async () => {
    const response = await fetch(endpoint);
    const result = await response.json();

    setData(result);
  };

  useEffect(() => {
    if (!endpoint) {
      return;
    }

    fetchData();
  }, [endpoint]);

  return data;
};

export const useFetchAll = urls => {
  const [responses, setResponses] = useState([]);

  useEffect(() => {
    const fetchData = async () => {
      const results = await Promise.all(
        urls.map(async url => {
          const response = await fetch(url);

          return response.json();
        })
      );

      setResponses(results);
    };

    if (urls.length) {
      fetchData();
    }
  }, [urls]);

  return responses;
};

export const useSearch = () => {
  const [value, setValue] = useState('');
  const [searchTerm, setSearchTerm] = useState('');
  const [searchFocus, setSearchFocus] = useState(false);
  const hasSearch = Boolean(searchFocus || searchTerm.length);

  const handleCancel = () => {
    setValue('');
    setSearchTerm('');
    setSearchFocus(false);
  };

  const handleChange = event => {
    setValue(event.target.value);
  };

  const handleFocus = () => {
    setSearchFocus(focus => !focus);
  };

  const searchHandlers = {
    onBlur: handleFocus,
    onCancel: handleCancel,
    onChange: handleChange,
    onFocus: handleFocus,
    value
  };

  useEffect(() => {
    if (value) {
      setSearchTerm(value.toUpperCase().trim());
    } else {
      setSearchTerm('');
    }
  }, [value]);

  return [searchTerm, searchHandlers, hasSearch];
};

export const useStockSearch = (allStocks, searchTerm) => {
  const [filteredStocks, setFilteredStocks] = useState(allStocks);

  useEffect(() => {
    if (searchTerm.length > 1) {
      const filtered = allStocks.filter(stock => {
        return (
          stock.symbol.indexOf(searchTerm) === 0 ||
          stock.description.indexOf(searchTerm) === 0
        );
      });

      setFilteredStocks(filtered);
    } else if (!searchTerm) {
      setFilteredStocks(allStocks);
    }
  }, [searchTerm]);

  return filteredStocks;
};

export const useStocks = initialStocks => {
  const [endpoints, setEndpoints] = useState([]);
  const [stocks, setStocks] = useState([]);
  const [allStocks, setAllStocks] = useState([]);
  const [activeStocks, setActiveStocks] = useState([]);
  const exchangeResponse = useFetch(API_EXCHANGE);
  const response = useFetchAll(endpoints);
  const cache = useStorageCache(
    'stocks',
    { stocks },
    stockData => !stockData.stocks.length
  );

  useEffect(() => {
    const cachedStocks = cache && cache.stocks;

    if (cachedStocks && cachedStocks.length) {
      setStocks(cachedStocks);
      setActiveStocks(cachedStocks);
    } else {
      setStocks(initialStocks);
      setActiveStocks(initialStocks);
    }
  }, []);

  useEffect(() => {
    if (!exchangeResponse) {
      // Exit early if no response
      return;
    }

    const symbols = activeStocks.map(stock => stock.symbol);
    const currentStocks = exchangeResponse.filter(stock =>
      symbols.includes(stock.symbol)
    );

    setEndpoints(currentStocks.map(getStockEndpoint));
    setAllStocks(exchangeResponse);
  }, [exchangeResponse]);

  useEffect(() => {
    if (activeStocks.length) {
      setEndpoints(activeStocks.map(getStockEndpoint));
    }
  }, [activeStocks]);

  useEffect(() => {
    if (!response.length || !activeStocks.length) {
      return;
    }

    const updatedStocks = response.map((result, index) => {
      const stock = activeStocks[index];

      return {
        ...result,
        ...stock,
        price: result.c,
        previousPrice: result.pc
      };
    });

    setStocks(updatedStocks);
  }, [response, activeStocks]);

  return [stocks, setStocks, allStocks];
};

export const useStockDetail = () => {
  const [endpoint, setEndpoint] = useState('');
  const [activeStock, setActiveStock] = useState(null);
  const [stock, setStockDetail] = useState(activeStock);
  const response = useFetch(endpoint);
  const getStock = symbol =>
    `https://finnhub.io/api/v1/quote?symbol=${symbol}&token=${API_TOKEN}`;

  useEffect(() => {
    setStockDetail(activeStock);

    if (!activeStock) {
      return;
    }

    if (!activeStock.price) {
      setEndpoint(getStock(activeStock.symbol));
    }
  }, [activeStock]);

  useEffect(() => {
    if (response) {
      setStockDetail({
        ...activeStock,
        ...response,
        price: response.c,
        previousPrice: response.pc
      });
      setEndpoint('');
    }
  }, [response]);

  return [stock, setActiveStock];
};

export const useStockNews = category => {
  const [endpoint, setEndpoint] = useState('');
  const [news, setNews] = useState([]);
  const response = useFetch(endpoint);
  const cache = useStorageCache(
    'stock-news',
    { news: response },
    res => !res.news.length
  );

  useEffect(() => {
    const cachedNews = cache && cache.news;

    if (cachedNews && cachedNews.length) {
      setNews(cachedNews);
      const expired = isExpiredNews(cachedNews[0].datetime);

      if (!expired && !category) {
        // Exit early when cache is not expired and no category
        return;
      }
    }

    setEndpoint(getStockNewsEndpoint(category));
  }, [category]);

  useEffect(() => {
    if (response && response.length) {
      setNews(response);
    }
  }, [response]);

  return news;
};

export const useStockChart = (symbol, range) => {
  const [endpoint, setEndpoint] = useState();
  const [updatedResponse, setResponse] = useState(null);
  const response = useFetch(endpoint);

  useEffect(() => {
    const updatedEndpoint = getCandlestickEndpoint(symbol, range);

    setEndpoint(updatedEndpoint);
  }, [symbol, range]);

  useEffect(() => {
    if (response && response.s === 'ok') {
      const data = updateChartData(response, range);

      setResponse(data);
    }
  }, [response]);

  return updatedResponse;
};

export const useStockTicker = (stocks, range = '1D') => {
  const [endpoints, setEndpoints] = useState([]);
  const [updatedResponse, setResponse] = useState([]);
  const response = useFetchAll(endpoints);
  const cache = useStorageCache(
    'stock-charts',
    { charts: updatedResponse },
    res => !res.charts.length
  );

  useEffect(() => {
    const cachedCharts = cache && cache.charts;

    if (cachedCharts && cachedCharts.length) {
      const updated = cachedCharts.map(rehydrateChartData);
      const [head] = updated;
      const expired = isExpired(head.chartData.endDate);

      setResponse(updated);

      if (!expired) {
        // Exit early when cache is not expired
        return;
      }
    }

    // Trigger fetch for candlestick data
    setEndpoints(getCandlestickEndpoints(stocks, range));
  }, [stocks.length]);

  useEffect(() => {
    if (response.length) {
      const updated = response.map((res, index) => {
        if (res.s !== 'ok') {
          return null;
        }

        const stock = stocks[index];
        const chartData = updateChartData(res, range);

        return {
          ...stock,
          chartData
        };
      });

      if (updated[0] !== null) {
        setResponse(updated);
      }
    }
  }, [response]);

  return updatedResponse;
};
