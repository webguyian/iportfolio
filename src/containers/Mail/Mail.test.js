import React from 'react';
import { create } from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';

import * as hooks from 'modules/mail/hooks';
import Mail from './Mail';

describe('<Mail />', () => {
  const useMail = hooks.useMail;

  hooks.useMail = jest.fn();

  const eventHandlers = {
    onCancel: jest.fn(),
    onConfirmCancel: jest.fn(),
    onDelete: jest.fn(),
    onSubmit: jest.fn()
  };

  beforeEach(() => {
    jest.clearAllMocks();
    // Reset to original implementation before each test
    hooks.useMail.mockImplementation(useMail);
  });

  it('renders correctly', () => {
    const component = create(
      <MemoryRouter>
        <Mail />
      </MemoryRouter>
    );

    expect(component).toMatchSnapshot();
  });

  it('renders correctly with controls', () => {
    const fields = {
      to: 'hello@webguyian.com',
      from: 'test@tester.com',
      subject: 'Testing...',
      body: 'This is a test'
    };
    hooks.useMail.mockReturnValueOnce([
      fields,
      jest.fn(),
      {
        invalid: false,
        showControls: true
      },
      eventHandlers
    ]);
    const component = create(
      <MemoryRouter>
        <Mail />
      </MemoryRouter>
    );

    expect(component).toMatchSnapshot();
  });
});
