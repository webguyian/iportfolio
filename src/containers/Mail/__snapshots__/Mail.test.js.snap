// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Mail /> renders correctly 1`] = `
<form
  className="mail-app"
  onSubmit={[Function]}
>
  <button
    aria-labelledby="mail-app"
    className="ui-btn mail-app-drag-handle"
    type="button"
  >
    <span
      className="ui-text ui-text--accessible"
    >
      Drag down to close
    </span>
  </button>
  <div
    className="mail-app-top-bar"
  >
    <Link
      onClick={[Function]}
      to="/home"
    >
      Cancel
    </Link>
  </div>
  <div
    className="mail-app-header"
  >
    <div
      className="mail-app-header-title"
    >
      <span
        className="ui-text ui-text--display"
        size="l"
      >
        New Message
      </span>
      <button
        className="ui-btn ui-btn--with-icon"
        disabled={false}
        type="submit"
      >
        <i
          className="ui-icon ui-icon-arrow-up fas fa-arrow-up fa-2x"
        />
        <span
          className="ui-text ui-text--accessible"
        >
          Send
        </span>
      </button>
    </div>
    <div
      className="mail-app-header-fields"
    >
      <div
        className="ui-form-field"
      >
        <label
          className="ui-form-field-label"
          htmlFor="to"
        >
          To:
        </label>
        <input
          autoComplete="off"
          className="ui-form-field-input"
          disabled={true}
          id="to"
          onChange={[Function]}
          readOnly={false}
          type="text"
          value="hello@webguyian.com"
        />
      </div>
      <div
        className="ui-form-field"
      >
        <label
          className="ui-form-field-label"
          htmlFor="from"
        >
          From:
        </label>
        <input
          autoComplete="off"
          className="ui-form-field-input"
          disabled={false}
          id="from"
          onChange={[Function]}
          placeholder="your@email.com"
          readOnly={false}
          type="email"
          value=""
        />
      </div>
      <div
        className="ui-form-field"
      >
        <label
          className="ui-form-field-label"
          htmlFor="subject"
        >
          Subject:
        </label>
        <input
          autoComplete="off"
          className="ui-form-field-input"
          disabled={false}
          id="subject"
          maxLength={35}
          onChange={[Function]}
          readOnly={false}
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
  <div
    className="ui-form-field mail-app-body"
  >
    <textarea
      className="ui-form-field-textarea"
      id="body"
      onChange={[Function]}
      value="

Sent from iPortfolio"
    />
  </div>
  <div
    className="mail-app-controls"
  >
    <button
      className="ui-btn ui-btn--anchor-block"
      onClick={[Function]}
      type="button"
    >
      Delete Draft
    </button>
    <Link
      className="ui-btn--anchor-block"
      to="/home"
    >
      Save Draft
    </Link>
    <button
      className="ui-btn ui-btn--anchor-block"
      onClick={[Function]}
      type="button"
    >
      Cancel
    </button>
  </div>
</form>
`;

exports[`<Mail /> renders correctly with controls 1`] = `
<form
  className="mail-app"
  onSubmit={[MockFunction]}
>
  <div
    className="mail-app-overlay"
  />
  <button
    aria-labelledby="mail-app"
    className="ui-btn mail-app-drag-handle"
    type="button"
  >
    <span
      className="ui-text ui-text--accessible"
    >
      Drag down to close
    </span>
  </button>
  <div
    className="mail-app-top-bar"
  >
    <Link
      onClick={[MockFunction]}
      to="/home"
    >
      Cancel
    </Link>
  </div>
  <div
    className="mail-app-header"
  >
    <div
      className="mail-app-header-title"
    >
      <span
        className="ui-text ui-text--display"
        size="l"
      >
        Testing...
      </span>
      <button
        className="ui-btn ui-btn--with-icon"
        disabled={false}
        type="submit"
      >
        <i
          className="ui-icon ui-icon-arrow-up fas fa-arrow-up fa-2x"
        />
        <span
          className="ui-text ui-text--accessible"
        >
          Send
        </span>
      </button>
    </div>
    <div
      className="mail-app-header-fields"
    >
      <div
        className="ui-form-field"
      >
        <label
          className="ui-form-field-label"
          htmlFor="to"
        >
          To:
        </label>
        <input
          autoComplete="off"
          className="ui-form-field-input"
          disabled={true}
          id="to"
          onChange={[Function]}
          readOnly={false}
          type="text"
          value="hello@webguyian.com"
        />
      </div>
      <div
        className="ui-form-field"
      >
        <label
          className="ui-form-field-label"
          htmlFor="from"
        >
          From:
        </label>
        <input
          autoComplete="off"
          className="ui-form-field-input"
          disabled={false}
          id="from"
          onChange={[Function]}
          placeholder="your@email.com"
          readOnly={false}
          type="email"
          value="test@tester.com"
        />
      </div>
      <div
        className="ui-form-field"
      >
        <label
          className="ui-form-field-label"
          htmlFor="subject"
        >
          Subject:
        </label>
        <input
          autoComplete="off"
          className="ui-form-field-input"
          disabled={false}
          id="subject"
          maxLength={35}
          onChange={[Function]}
          readOnly={false}
          type="text"
          value="Testing..."
        />
      </div>
    </div>
  </div>
  <div
    className="ui-form-field mail-app-body"
  >
    <textarea
      className="ui-form-field-textarea"
      id="body"
      onChange={[Function]}
      value="This is a test"
    />
  </div>
  <div
    className="mail-app-controls mail-app-controls--show"
  >
    <button
      className="ui-btn ui-btn--anchor-block"
      onClick={[MockFunction]}
      type="button"
    >
      Delete Draft
    </button>
    <Link
      className="ui-btn--anchor-block"
      to="/home"
    >
      Save Draft
    </Link>
    <button
      className="ui-btn ui-btn--anchor-block"
      onClick={[MockFunction]}
      type="button"
    >
      Cancel
    </button>
  </div>
</form>
`;
